### ISSUE TITLE
"Easier" installation?

### ISSUE THREAD

**Issue by alphapapa** [raxod502/straight.el#302]
Hi again,

I was thinking about giving straight.el a try for installing some of my own packages (rather than waiting for MELPA to rebuild them, or manually copying them into my Emacs config when I update them), but reading the installation instructions is...intimidating, and a bit unsettling.  :)  For example, I recoil when told to eval code that downloads and evals code directly from `raw.githubusercontent.com`.  (Sure, I'm still trusting GitHub when I clone a git repo, but the point remains that, in principle, it's an anti-pattern, almost as bad as `curl | bash`.)

In contrast, setting up package.el in my config is as simple as:
```el
(require 'package)
(setq package-archives '(("gnu" . "https://elpa.gnu.org/packages/")
                         ("melpa" . "https://melpa.org/packages/")))
(package-initialize)
```

I suppose that, ideally, installing straight would be as simple as copying some files into my `load-path` and then doing something like `(require 'straight) (straight-init)`.  Even better would be if I could install and initialize straight *with* package.el, but I'm sure *you* recoil at that idea, haha.  ;)

Anyway, my question is, is anything like that on your radar, or is this install.el->bootstrap.el process how straight is always going to be installed and activated?

Thanks.

**Comment by raxod502** [comment]
If you don't like the bootstrap installation process, then you can just clone `straight.el` to `~/.emacs.d/straight/repos/straight.el` and include

```elisp
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (load bootstrap-file nil 'nomessage))
```

in your init-file. As you may notice, the installation process only happens if `straight.el` is not already installed.

The reason I don't recommend this is that you have to manually select the correct revision and branch and fork of `straight.el`, if any of those have been customized, whereas the automatic installation does all of this automagically.

---

Personally I'm not sure why you would find this easier than just pasting some code into your init-file, but I can certainly understand that it's safer, in that the initial bootstrap process is the only time when we run code not from a Git repository.

The reason `package.el` _looks_ simpler is because the code you quoted is the analogue of the code I quoted above: it only _initializes_ `package.el`; it doesn't _install_ it. When you take that into account, the `straight.el` initialization code is of equal complexity.

It's clearly impossible to install `straight.el` correctly with `package.el`, given how limited `package.el` is. It would only be possible to install a stub which then installs `straight.el` or directs the user on how to do so. This can't be registered on MELPA, however, as per https://github.com/melpa/melpa/issues/4939, and it's almost certainly never going to be on GNU ELPA for ideological reasons, so at that point I'd be maintaining a custom ELPA repository and this all gets even more ridiculously convoluted than `install.el`.

At the end of the day, I think it only makes sense for you to do it all manually, or to let `straight.el` do it correctly and automatically for you. There's not really a middle ground, since I don't want to implement some mode of operation that's automatic but operates incorrectly in some cases like when you specify a custom fork for `straight.el`. Why not just make the automatic installer more secure, if that's the primary concern?

**Comment by raxod502** [comment]
I think there may have been a miscommunication somewhere. You seem to think that the `straight.el` installer modifies the init-file, which is not the case. The _only_ thing the installer does is clone the `straight.el` repository to `~/.emacs.d/straight/repos/straight.el`.

You should think of it like so:

```elisp
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-magic-so-that-this-works-correctly
      (straight-use-package 'straight)))
  (load bootstrap-file nil 'nomessage))
```

And I really do mean that. The reason for `install.el` is to carefully support _exactly_ the same interface for the installation of `straight.el` as is supported for the installation of any other package. In particular, if a particular version is specified in the user's lockfiles, then that revision is installed; if a custom branch is specified, then it's checked out; if the user overrides the recipe to some other value, then all possible customizations for that are fully supported. This includes, in particular, switching to your own fork of `straight.el`. Why should switching to your fork of `straight.el` be any harder than switching to your fork of any other package? It shouldn't—and isn't, because we have the installer.

In other words, I think you are analogizing the `straight.el` installer to something like [the zplugin installer](https://github.com/zdharma/zplugin/blob/b77426f6357f69475333dc27c1edc8324e05b5e7/doc/install.sh#L33-L42), which is not a good analogy. Usually, `curl | sh` installers

1. modify some user configuration file, which is bad
2. have no particular reason for existence other than convenience, which is also bad

Neither of these two conditions is true for `straight.el`, since the installer does not modify any user configuration file, and it provides support for important features that are impossible to provide any other way.

Does that change your mind about using the `straight.el` installer?

> It might be good to mention that in the docs.

I will consider that, but first I want to hear your thoughts on whether ordinary users would have any interest in this mode of operation, and if so why—given what I've said above.

> What I really want in my Emacs config is consistency, and for packages to not change out from under me.

Well, yes. This is the entire point of the installer. It installs exactly the version of `straight.el` that is specified in your lockfile. It does _not_ install the latest version. If you install manually, then you have to manually make sure to check out the correct version. So I'm not sure how the installer could cause a loss in consistency.

Are you perhaps confused by the fact that the URL referenced in the installer points to a file on the latest `develop`? If you have a lockfile, then that revision of `straight.el` is the one installed. Furthermore, bugs on the `develop` branch of `straight.el` do not affect the installation process. The `install.el` module is designed to be very stable and to work for _all_ versions of `straight.el`, past and present. I have the installer always request the latest version so that I can backport fixes to critical bugs in the installation process.

> Besides that, I don't want "automagic" stuff in my Emacs config.

I have a similar sentiment, but I don't agree with your conclusion. Personally, what _I_ find distasteful is when an automagical tool fails to provide a clean API. Magit works perfectly out of the box, and I have no idea what's going on inside, but there's an obvious relationship between my input keypresses and the resulting Git operations. Similarly the `straight.el` installer provides an easily understandable interface to installing `straight.el` (namely exactly the same one that is provided for installing any other package). The only time I object to magic (for example, in most Apple products) is when the interface is not clearly specified or the abstractions are leaky—at which point the magic becomes a disadvantage in debugging.

> Not knowing everything that straight does, I don't see why that should be impossible

Firstly, `package.el` would need to support cloning a Git repository. It doesn't. Secondly, `package.el` would need to support parsing `straight.el` lockfiles. It doesn't. Thirdly, `package.el` would need to support installing an arbitrary revision of a package from an arbitrary source dynamically determined at runtime without prior configuration of an ELPA repository. Not even close.

Just look at `install.el`—that's what's necessary in order to correctly install `straight.el`. The only way to do that with `package.el` is to use `package.el` to install `install.el`, which then installs `straight.el`. You could do that, but as said earlier, it would just be even more convoluted than downloading `install.el` directly. (It would also be slower.)

> But that isn't your design goal, which is fine.

I didn't say that. In fact, as I said above, I explicitly [asked the MELPA maintainers if I could do this](https://github.com/melpa/melpa/issues/4939) and they said no.

> I just have an aversion to running code that modifies my init file. I would trust straight.el to modify files in its subdirectory.

Running the installer only modifies files in `~/.emacs.d/straight`.

> I don't understand why straight.el needs to be aware of its own provenance.

Because `straight.el` treats itself as exactly equivalent to any other package, in all situations.

> Unless, of course, it's designed to upgrade itself, which is also a cool feature

Of course it supports this. A package manager which can't upgrade itself would be pretty dumb in my opinion. I also feel strongly that upgrading forks transparently is a must-have feature, given the philosophy of `straight.el` (which is to make forks and pull requests of Emacs packages trivial to manage).

> a general aversion to code modifying my init file (yes, I know package.el does that when necessary, but it is included with Emacs, and I added the code myself anyway)

Note that `straight.el` _never_ modifies any user configuration file, _unlike_ `package.el`. This is because I am even more averse to automatic init-file modification than you, to the point that `straight.el` specifically includes code that disables that `package.el` feature automatically.

> I don't think there's much that can be done about that, unless you implement PGP signatures

Right. I think one simply must rely upon HTTPS, and the paranoid people would benefit from a mode of operation where `straight.el` refuses to activate a package until that particular revision is audited by the user. We simply don't have PGP signatures for all GitHub repositories, after all. See https://github.com/raxod502/straight.el/issues/85 and initial discussion in https://github.com/raxod502/straight.el/issues/159.

---

I am gathering that the current documentation does not communicate this information well. It is difficult for me to see what should be changed, since I have an understanding of all the concepts already, so suggestions for improvement would be greatly appreciated.

**Comment by raxod502** [comment]
This issue is being closed automatically since it has a "waiting on response" label but hasn't had any activity in 90 days. But don't worry: if you have some information to help move the discussion along, just leave a comment and I'll re-open the issue.

*(Comment generated automatically by [NO CARRIER](https://github.com/twbs/no-carrier) via [tidier](https://github.com/raxod502/tidier).)*

**Comment by progfolio** [comment]
I do think this is worth reconsidering at some point.
`install.el` requires that we maintain some duplicate code in two separate files:
https://github.com/raxod502/straight.el/issues/418#issuecomment-695275251

I realize this is easier said than done. I'll think a bit on this and keep you posted if I get any ideas.

**Comment by yuuyins** [comment]
@raxod502 @progfolio can you at least upload the public key? gpg complains about the ones from keys.openpgp.org and keys.keyserver.ubuntu.com

gpg: key 4AEE18F83AFDEB23: new key but contains no user ID - skipped

**Comment by raxod502** [comment]
Public key? There is no GPG signing anywhere in the build or distribution process of `straight.el`.

**Comment by yuuyins** [comment]
@raxod502 commit signing aka verified. 
- https://github.com/radian-software/straight.el/commit/c64a1f668d758e06f15d8ba8f5cfc81273a45f20
- https://github.com/radian-software/straight.el/commit/94303163f8c72f6325e36d88dc0c27003619fc14

**Comment by raxod502** [comment]
Ah. GitHub adds my GPG key automatically to commits via the web UI, it seems. You can disregard that, other commits are not signed.

**Comment by guest936723** [comment]
Wanted to jump in and add my 2¢. First, thank you for the care and thought you all put into this. It was nice to see all these points already discussed.

Second, I think one thing that hasn't been said yet that could be useful is: if GPG verification were to be added for `install.el` (e.g. pulling a pub key in the bootstrap snippet & verifying the file before using it), this would remove the need to trust Github, which would make me feel better here.

This is slightly different than `curl | bash` in (e.g.) rustup, where you're trusting the domain owners _and_ HTTPS (which seems ok), but in this case we're unnecessarily trusting a 3rd party (Github).

**Comment by raxod502** [comment]
That's a fair suggestion but I'm not sure it actually adds much security to implement. You can verify the bootstrap snippet with an out-of-band PGP signature all you want, but then you still have to clone the actual source code for `straight.el` from GitHub, not to mention every other package you use, so you're still trusting GitHub.

**Comment by guest936723** [comment]
Good point. Another thought is to use Emacs' built-in package signing... but I'm guessing that's only relevant if we use package.el or parts of it? I'm not familiar enough with the space to make a coherent suggestion, just want to reduce the amount of trust we need to place in 3rd parties if possible :).

Unfortunate MELPA doesn't want meta-packages too.

**Comment by raxod502** [comment]
Package signing doesn't really work for building packages from source, which is a cornerstone of the design of `straight.el`. Unfortunately, I don't think there's much that can be done here.

**Comment by raxod502** [comment]
I don't think there's much to be done here. As discussed above, the bootstrap snippet is optional, and this fact is now documented in the installation instructions for `straight.el`. If you don't want to use it, you can just clone the repo manually as described there.

However, cloning Git repositories for other packages is a fundamental aspect of the design of `straight.el` and changing that is a much different discussion better suited to another issue report with a concrete idea about what could be done differently.

Closing as completed since the question in the title is answered with a yes.
